SmartDeviceLink/SDLCacheFileManager.m

```
@@ -0,0 +1,282 @@
+*/
+- (nullable NSString *)sdl_writeImage:(UIImage *)icon toFileFromURL:(NSString *)urlString {
+    NSData *iconPNGData = UIImagePNGRepresentation(icon);
+    NSString *iconStorageName = [self.class sdl_md5HashFromString:urlString];
+    
+    NSString *imageFilePath = [self.sdl_cacheFileBaseDirectory stringByAppendingPathComponent:iconStorageName];
+    [iconPNGData writeToFile:imageFilePath atomically:YES];
+    
+    return imageFilePath;
+}
+
+/**
+ * Handles request to update archive file. Will create a new archive file if one is not present.
+ *
+ * @param request The System request used to retrieve the icon
+ * @param iconFilePath The directory path where the icon file is held
+ * @param archiveFile Current archive file to update
+*/
+- (void)updateArchiveFileFromRequest:(SDLOnSystemRequest *)request iconFilePath:(NSString *)iconFilePath archiveFile:(SDLIconArchiveFile *)archiveFile {
+    NSError *error = nil;
+    NSMutableArray *archiveArray = [NSMutableArray new];
+    
+    if (archiveFile.lockScreenIconCaches != nil) {
+        archiveArray = [archiveFile.lockScreenIconCaches mutableCopy];
+    }
+    
+    // Need to remove duplicate SDLLockScreenIconCache object if handling expired icon
+    for (SDLLockScreenIconCache *iconCacheCopy in archiveArray) {
+        if ([iconCacheCopy.iconUrl isEqualToString:request.url]) {
+            [archiveArray removeObject:iconCacheCopy];
+            break;
+        }
+    }
+    
+    SDLLockScreenIconCache *newIconCache = [[SDLLockScreenIconCache alloc] initWithIconUrl:request.url iconFilePath:iconFilePath];
+    [archiveArray addObject:newIconCache];
+    
+    archiveFile.lockScreenIconCaches = archiveArray;
+    
+    // HAX: Update this when we are iOS 11.0+
+    BOOL writeSuccess = [NSKeyedArchiver archiveRootObject:archiveFile toFile:self.sdl_archiveFileDirectory];
+    if (!writeSuccess) {
+        SDLLogE(@"Could not write file to specified path: %@", error);
```

This should return an error if it failed

---

Just leaving as a note, per our previous discussions with methods being used within `retrieveImageForRequest:withCompletionHandler`, logging the error should be sufficient here as we will still return the image in the retrieve method should this fail. Let me know if I am mistaken.

---

I don't think so. This is a "sub" method, not the primary logic. Updating the archive file failed, so it should return an error, right? The main logic can decide what to do with that (and it should continue to download the file), but this method failed.